### PR TITLE
Add openstreet service

### PR DESCRIPTION
**Pull request by samueljohnsegun148**
Please ensure you've followed the checklist and provide all the required information *before* requesting a review.

## Required Information

- [x] Reference the issue this is meant to fix or describe it if none exists.
- [x] Full description of changes made.

## Pull Request Checklist

- [x] Coding standards are followed (e.g., [PEP8](https://pep8.org/))
- [x] An entry exists for the container in `docker-compose.yml` or `build.yml`
- [x] The Docker image builds
- [x] The container runs correctly when sent inputs with the changes
- [x] No models or other large files are part of these commits
- [x] A description of the tests already run as part of this PR is present

### CODE REVIEW

**Comment by jeffbl** [comment]
@samueljohnsegun148 I believe this is a skeleton checkin of a service for OSM data? If so, this should not be in the "services" directory, but rather in the "preprocessors" directory, since it would be a preprocessor. To merge this branch into main, we would need all of the above boxes checked off, and details on what testing has been done, etc. so that it can be sent to @jaydeepsingh25 for formal PR review and merge. Until it has functionality that is working and tested, it should remain in its own branch. When complete, it should reference the work items that the PR resolves.

@BenMacnaughton @aidanwilliams09 Can you please schedule a time with @samueljohnsegun148 to go over the process, and help get the OSM preprocessor moving forward?

Assigning to @samueljohnsegun148 and marking as a draft PR.

**Comment by samueljohnsegun148** [comment]
> @samueljohnsegun148 I believe this is a skeleton checkin of a service for OSM data? If so, this should not be in the "services" directory, but rather in the "preprocessors" directory, since it would be a preprocessor. To merge this branch into main, we would need all of the above boxes checked off, and details on what testing has been done, etc. so that it can be sent to @jaydeepsingh25 for formal PR review and merge. Until it has functionality that is working and tested, it should remain in its own branch. When complete, it should reference the work items that the PR resolves.
> 
> @BenMacnaughton @aidanwilliams09 Can you please schedule a time with @samueljohnsegun148 to go over the process, and help get the OSM preprocessor moving forward?
> 
> Assigning to @samueljohnsegun148 and marking as a draft PR.

**Comment by samueljohnsegun148** [comment]
@jeffbl . yes it is to test the service of OSM data.

**Comment by jeffbl** [comment]
Deleted comment above that was just a quote. And, I misspoke in my comment above calling it the "OSM service". It should be "OSM Preprocessor" (corrected).

**Comment by aidanwilliams09** [comment]
In order to check off the remaining boxes:

First and foremost, for the docker-compose.yml, you are going to want to move both your preprocessor and service docker-compose.yml into the main docker-compose at the directory root. Additionally, I'd recommend following the format that all other entries in that file follow. 

For the Docker image building, you'll want to access unicorn and pull this branch and make sure you have the latest schemas. After, you will want to set the $DOCKER_GID environment variable to 134, change the docker-compose image name for the preprocessor and handler from ghcr... to osm_preprocessor:latest  and osm_handler:latest to ensure you do a local build, and then run the commands:
```bash
docker build schemas -t schemas
docker-compose -f build.yml build <service>
docker-compose up -d orchestrator <service>
```
where <service> is the name of the entry of your preprocessor/handler in docker-compose.yml.
If there are no errors and you can successfully run the preprocessor and handler then that step should be checked off.

Lastly, for testing, I cannot really help too much. You will want to run the images and send requests to the server to ensure it works as you expect but beyond that you will have to decide what tests are important and then document those tests as comments here.

**Comment by aidanwilliams09** [comment]
And for the Required Information checklist, link the issue in the development area on the right side of the PR, I want to say this is IMAGE-server#205 but please confirm this yourself. And then provide a comment here describing each portion of your PR, especially because this follows a fairly different format than all other preprocessors and handlers.

File: docker-compose.yml
Comment (aidanwilliams09) [inline docker-compose.yml:100]: Change this to the ghcr one for production. "osm-preprocessors:latest" should only be used for your local testing on unicorn

**Comment by samueljohnsegun148** [comment]
### Instructions for testing

1. Clone this repository. Note that the schemas are a submodule, so you need to either get them in the initial clone, e.g.,
```
git clone --recurse-submodules git@github.com:Shared-Reality-Lab/IMAGE-server.git
```

or else get them after you've done the initial clone (while in the root of the cloned repo on your local machine):
```
git submodule init
git submodule update

```

2. Run 

```
git checkout openStreetM

```
3. Set DOCKER_GID variable to 134

```
export DOCKER_GID=134

```

4. Run

```
docker build schemas -t schemas
docker-compose build openstreet
docker-compose up -d orchestrator openstreet
docker run -d osm-preprocessors

```

5. Test with the request below to get sample result
```
curl -X 'GET' \
  'http://localhost:8000/location/100/49.8974309/-97.2033944' \
  -H 'accept: application/json'

```
In this case, 100 is the distance in metres while 49.8974309 and -97.2033944 are
latitude and longitude respectively of a point on the OpenStreet Map. You may play around the three values to see different results yourself.



You may also do the testing on your local machine by following all the above instructions. 
On your local machine, you may also like to download and install insomnia rest. Please see https://insomnia.rest/    

After the installation, type http://localhost:8000/location/100/49.8974309/-97.2033944 on the "GET" field.
and click the "Send Button". You may also like to use postman if you so wish. The displayed results should be the same as that of (5) above but in a more user-friendly format.

File: preprocessors/openstreet/app/main.py
Comment (JRegimbal) [inline preprocessors/openstreet/app/main.py:83]: I'm a bit confused, why are you returning comments in the data like this instead of using a key-value pair?

File: preprocessors/openstreet/app/osm_service.py
Comment (JRegimbal) [inline preprocessors/openstreet/app/osm_service.py:8]: Top level description should be provided for these functions with general information (e.g., how should these values be conceptually understood to relate to the resulting bounding box *without* stepping through the rest of the code).

**Review by JRegimbal — CHANGES_REQUESTED** [review]
Empty config files and IMAGE-server submodule need to be deleted. Submodule doesn't make sense in here and the added configs should be part of your local gitignore if necessary.

Note that I'm not really looking at the main.py/osm_service.py right now since they're not compatible with the [POST /preprocessor format](https://github.com/Shared-Reality-Lab/IMAGE-server/wiki/2.-Handlers,-Preprocessors-and-Services#preprocessors=) necessary for integration with IMAGE. Once this is done and working, I'll take a closer look here.

**Comment by samueljohnsegun148** [comment]
Thank you Juliette for the corrections.
I am sorry, my system developed fault, so I'm looking for the alternative
at the moment.

However, those issues raised will be addressed shortly.

Thanks once again.

Regards,
Segun

File: .github/workflows/openstreet.yml
Comment (samueljohnsegun148) [reply .github/workflows/openstreet.yml]: The E501 error were mostly from the virtual environment files, but has been addressed now.

File: docker-compose.yml
Comment (samueljohnsegun148) [reply docker-compose.yml]: The service name has now been changed from "openstreet" to "openstreetmap"

File: preprocessors/openstreet/app/main.py
Comment (samueljohnsegun148) [reply preprocessors/openstreet/app/main.py]: Has now been addressed

File: preprocessors/openstreet/app/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreet/app/osm_service.py]: Top level description is now in the README file.  Also look at this: https://docs.google.com/document/d/1RN9tQKsTodWhX7qpr40jLGHbjzhyM4RrG0JYlSvcMrw/edit?usp=sharing

File: preprocessors/openstreet/docker-compose.yml
Comment (samueljohnsegun148) [reply preprocessors/openstreet/docker-compose.yml]: It was for local testing. It is no longer committed

File: preprocessors/openstreet/app/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreet/app/osm_service.py]: Top level description now in the README file. You can also check: https://docs.google.com/document/d/1RN9tQKsTodWhX7qpr40jLGHbjzhyM4RrG0JYlSvcMrw/edit?usp=sharing

File: preprocessors/openstreet/Readme.md
Comment (samueljohnsegun148) [reply preprocessors/openstreet/Readme.md]: Noted

File: docker-compose.yml
Comment (JRegimbal) [inline docker-compose.yml:15]: These lines should not have been changed.

File: docker-compose.yml
Comment (JRegimbal) [reply docker-compose.yml:69]: There's still whitespace.

File: docker-compose.yml
Comment (JRegimbal) [inline docker-compose.yml:80]: Remove these commented out lines.

File: preprocessors/openstreetmap/Dockerfile
Comment (JRegimbal) [inline preprocessors/openstreetmap/Dockerfile:29]: Remove commented out lines.

File: preprocessors/openstreetmap/main.py
Comment (JRegimbal) [inline preprocessors/openstreetmap/main.py:27]: Do you actually respond meaningfully to a GET request here?

File: preprocessors/openstreetmap/main.py
Comment (JRegimbal) [inline preprocessors/openstreetmap/main.py:111]:
```suggestion
    app.run(host='0.0.0.0', port=5000, debug=False)
```

File: preprocessors/openstreetmap/osm_service.py
Comment (JRegimbal) [inline preprocessors/openstreetmap/osm_service.py:184]: Overpass API is part of OSM?

**Review by JRegimbal — CHANGES_REQUESTED** [review]
Very long PR, but looking good. Some comments made but I think it should be ready once the schemas PR is merged, the schemas submodule here is updated appropriately, and everything is tested in the architecture on Unicorn (i.e., I should be able to see the OSM output on a request to `/render/preprocess`).

File: preprocessors/openstreetmap/main.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/main.py]: It is possible to make it respond to a GET request as well, if I chose to. However, this is not the primary focus here as I only intend to get it respond to a POST request, so I have deleted the GET method.

File: preprocessors/openstreetmap/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:182]: Yes, OSM uses Overpass API.
https://wiki.openstreetmap.org/wiki/Overpass_API

**Review by JRegimbal — CHANGES_REQUESTED** [review]
I'm a bit confused about the utility of `n/a` as a response for street information. Is there a particular reason this is preferred as to omitting that key? My expectation would be that if, for example, a street name was not available, the field would not be present rather than it containing a string that, unless you are explicitly checking for "n/a", would otherwise be assumed as valid. The same goes for other fields like "oneway" which I assume would typically be a boolean but is encoded as a string. Are these behaviors that Florian requested?

The main issue though is speed and reliability. In my testing on Unicorn with the following dummy request, the preprocessor will often timeout, failing to send a response within the allowed 15 s. There also seems to be a possibly related issue with the API in use with us running out of requests. See the error message below. Note that this too many requests error does not always occur when the preprocessor fails to respond quickly enough.
```
openstreetmap_1                | ERROR:main:Exception on /preprocessor [POST]
openstreetmap_1                | Traceback (most recent call last):
openstreetmap_1                |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2077, in wsgi_app
openstreetmap_1                |     response = self.full_dispatch_request()
openstreetmap_1                |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1525, in full_dispatch_request
openstreetmap_1                |     rv = self.handle_user_exception(e)
openstreetmap_1                |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1523, in full_dispatch_request
openstreetmap_1                |     rv = self.dispatch_request()
openstreetmap_1                |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1509, in dispatch_request
openstreetmap_1                |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
openstreetmap_1                |   File "/app/main.py", line 80, in get_map_data
openstreetmap_1                |     amenity = get_amenities(bbox_coordinates)
openstreetmap_1                |   File "/app/osm_service.py", line 187, in get_amenities
openstreetmap_1                |     amenities = api.query(
openstreetmap_1                |   File "/usr/local/lib/python3.9/site-packages/overpy/__init__.py", line 180, in query
openstreetmap_1                |     raise current_exception
openstreetmap_1                | overpy.exception.OverpassTooManyRequests: Too many requests
openstreetmap_1                | [2022-06-13 14:44:43 +0000] [7] [DEBUG] Ignoring EPIPE
```

**Comment by JRegimbal** [comment]
Apologies, forgot to include the dummy request.
```json
{
  "request_uuid": "a3d7b6be-8f3b-47ba-8b17-4b0557e7a8ce",
  "timestamp": 1637789517,
  "coordinates": {
      "latitude": 47.620452,
      "longitude": -122.348793
  },
  "context": "",
  "language": "en",
  "url": "https://fake.site.com/some-url",
  "capabilities": [],
  "renderers": [
    "ca.mcgill.a11y.image.renderer.Text",
    "ca.mcgill.a11y.image.renderer.SimpleAudio",
    "ca.mcgill.a11y.image.renderer.SegmentAudio"
  ]
}
```

**Comment by samueljohnsegun148** [comment]
> overpy.exception.OverpassTooManyRequests: Too many requests



> I'm a bit confused about the utility of `n/a` as a response for street information. Is there a particular reason this is preferred as to omitting that key? My expectation would be that if, for example, a street name was not available, the field would not be present rather than it containing a string that, unless you are explicitly checking for "n/a", would otherwise be assumed as valid. The same goes for other fields like "oneway" which I assume would typically be a boolean but is encoded as a string. Are these behaviors that Florian requested?

"n/a" comes in when the information for the given key is not available from OSM. However, I can work around it to automatically remove empty keys as suggested, and it will be effected shortly.

**Comment by samueljohnsegun148** [comment]
> The main issue though is speed and reliability. In my testing on Unicorn with the following dummy request, the preprocessor will often timeout, failing to send a response within the allowed 15 s. There also seems to be a possibly related issue with the API in use with us running out of requests. See the error message below. Note that this too many requests error does not always occur when the preprocessor fails to respond quickly enough

I will figure out why the speed is that low and how it can be improved, as it seems unusual to me. As for the "overpy.exception.OverpassTooManyRequests: Too many requests" error, it may occur at times when Overpass servers experience heavy load. So, with a retry option, it should work. But I am equally looking at ways to address this better.

**Comment by JRegimbal** [comment]
Are the performance issues fixed?

**Comment by samueljohnsegun148** [comment]
> Apologies, forgot to include the dummy request.
> 
> ```json
> {
>   "request_uuid": "a3d7b6be-8f3b-47ba-8b17-4b0557e7a8ce",
>   "timestamp": 1637789517,
>   "coordinates": {
>       "latitude": 47.620452,
>       "longitude": -122.348793
>   },
>   "context": "",
>   "language": "en",
>   "url": "https://fake.site.com/some-url",
>   "capabilities": [],
>   "renderers": [
>     "ca.mcgill.a11y.image.renderer.Text",
>     "ca.mcgill.a11y.image.renderer.SimpleAudio",
>     "ca.mcgill.a11y.image.renderer.SegmentAudio"
>   ]
> }
> ```



> Are the performance issues fixed?

Yes, it seems to be okay. Could you please try it again from your end?  
Most public overpass servers have a rate limit. The main overpass server is often heavily used and may result at times in rate limit (although they indicate that one can safely query up to 10000 times per time, which is mostly not the case),- https://dev.overpass-api.de/overpass-doc/en/preface/commons.html. 

I have also tried to further limit the amount of data returned by the queries, to largely reduce the likelihood of the preprocessor not responding within the 15 seconds requirement. I think the longest time ever from my end is 10 seconds.

**Comment by samueljohnsegun148** [comment]
> Apologies, forgot to include the dummy request.

Kindly let me have your observations upon review.

File: .gitignore
Comment (JRegimbal) [inline .gitignore:5]: This should not have been changed. These files should *not* be committed.

**Review by JRegimbal — CHANGES_REQUESTED** [review]
Changes to .gitignore and the commit of blank config files should be reversed. This will likely cause problems later on.

With respect to the performance updates, I can't make the preprocessor time out due to a long request. However, the Overpass API server still will return errors including the too many requests error and a gateway timeout error (which will cause the preprocessor to time out). Since these API requests fail so frequently, I'd recommend wrapping properly handling those errors when thrown. I also worry about deploying a preprocessor reliant on such an unreliable service into production. Have you looked into the possibility of a self-hosted instance? If the resource requirements aren't too high, that may be a better approach and then we could set the server used as an environment variable of the container at runtime.

**Comment by JRegimbal** [comment]
Here's a quick summary of the tests I've been doing:

I've spun up an orchestrator instance and an openstreetmap instance on Unicorn on their own. I'll then send the contents of the JSON file above to the orchestrator under the `/render/preprocess` path. This ends up looking like `sendiagereq $PATH_TO_JSON http://$ORCHESTRATOR_IP:8080/render/preprocess` or equivalently `curl -H "Content-Type: application/json" -d@$PATH_TO_JSON http://$ORCHESTRATOR_IP:8080/render/preprocess`. I monitor the logs of these containers using `docker-compose logs -f`. The 15 s timeout is a limit enforced by the orchestrator to prevent preprocessors from delaying the message. Timeouts between the openstreetmap container and the Overpass API server only happened once with a gateway timeout error returned by the preprocessor *after* the orchestrator had already timed out waiting on the preprocessor.

File: docker-compose.override.yml
Comment (samueljohnsegun148) [reply docker-compose.override.yml:1]: > Just to let you know, you shouldn't ever need to commit an override file for testing purposes and you only need to replace the keys that have changed from the main `docker-compose.yml`. That is, the `labels` could likely be omitted.

Okay, thanks. 



> Here's a quick summary of the tests I've been doing:
> 
> I've spun up an orchestrator instance and an openstreetmap instance on Unicorn on their own. I'll then send the contents of the JSON file above to the orchestrator under the `/render/preprocess` path. This ends up looking like `sendiagereq $PATH_TO_JSON http://$ORCHESTRATOR_IP:8080/render/preprocess` or equivalently `curl -H "Content-Type: application/json" -d@$PATH_TO_JSON http://$ORCHESTRATOR_IP:8080/render/preprocess`. I monitor the logs of these containers using `docker-compose logs -f`. The 15 s timeout is a limit enforced by the orchestrator to prevent preprocessors from delaying the message. Timeouts between the openstreetmap container and the Overpass API server only happened once with a gateway timeout error returned by the preprocessor _after_ the orchestrator had already timed out waiting on the preprocessor.

I stand to be corrected. It looks  to me like I'm employing an entirely different approach for the POST request, on my local terminal using Git Bash. On the unicorn server, I encountered an error when running  `docker-compose up -d orchestrator openstreetmap`.  I precisely got this: 
"ERROR: for orchestrator  Cannot start service orchestrator: Unable to find group
ERROR: Encountered errors while bringing up the project."
Do I need permissions to run it? I guess I would need the orchestrator to work well before I can access it for a request via ` /render/preprocess`.

More also, it seems to me that the schema for the openstreetmap is yet to be merged to the main, or is there any thing I'm not getting well? I received this error: "
 File "/app/main.py", line 34, in get_map_data
    with open('./schemas/preprocessors/openstreetmap.schema.json') as jsonfile:
FileNotFoundError: [Errno 2] No such file or directory: './schemas/preprocessors/openstreetmap.schema.json'
^C[2022-06-15 12:16:11 +0000] [1] [INFO] Handling signal: int
[2022-06-15 12:16:11 +0000] [7] [INFO] Worker exiting (pid: 7)
[2022-06-15 12:16:11 +0000] [1] [INFO] Shutting down: Master"

The above error came after I sent the request below directly on the unicorn server:
``` curl -X POST \
  http://localhost:5000/preprocessor \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -H 'postman-token: 0bc130ca-ab92-ba34-bb65-603e5b0acb82' \
  -d '{
  "request_uuid": "a3d7b6be-8f3b-47ba-8b17-4b0557e7a8ce",
  "timestamp": 1637789517,
  "coordinates": {
      "latitude": 47.620452,
      "longitude": -122.348793
  },
  "context": "",
  "language": "en",
  "url": "https://fake.site.com/some-url",
  "capabilities": [],
  "renderers": [
    "ca.mcgill.a11y.image.renderer.Text",
    "ca.mcgill.a11y.image.renderer.SimpleAudio",
    "ca.mcgill.a11y.image.renderer.SegmentAudio"
  ]
}'

```

Comment (JRegimbal) [reply docker-compose.override.yml:1]: You would need to update the submodule to current schemas. Main is currently set up to use a slightly older version. It should be updated as part of your PR. Rebuilding the openstreetmap container should fix this. You can use the remote version of the orchestrator as well.

As for the group error with the orchestrator, you need to specify the docker group ID to it in the environment variable `DOCKER_GID`. On Unicorn, this would be 134, but note that this value will change from system to system.

**Comment by samueljohnsegun148** [comment]
> @samueljohnsegun148 please revert [f8dd310](https://github.com/Shared-Reality-Lab/IMAGE-server/commit/f8dd3100f214f2cf7fb1efd51f7ea73ab3bdcbc0) and [4fc3be7](https://github.com/Shared-Reality-Lab/IMAGE-server/commit/4fc3be7e89c503666133a72914382e1be0ec7381). These changes should not have been pushed.
> 
> Actually, all of these recent changes to the orchestrator in docker-compose.yml should not have been pushed. These will need to be fixed before merge.

Yes, for sure. The changes have been undone.

**Comment by samueljohnsegun148** [comment]
> Changes to .gitignore and the commit of blank config files should be reversed. This will likely cause problems later on.
> 
> With respect to the performance updates, I can't make the preprocessor time out due to a long request. However, the Overpass API server still will return errors including the too many requests error and a gateway timeout error (which will cause the preprocessor to time out). Since these API requests fail so frequently, I'd recommend properly handling those errors when thrown. I also worry about deploying a preprocessor reliant on such an unreliable service into production. Have you looked into the possibility of a self-hosted instance? If the resource requirements aren't too high, that may be a better approach and then we could set the server used as an environment variable of the container at runtime.

Owning an Instance of Overpass API should address this:
Requirements: 
•	POSIX conforming operating system such as Linux, FREEBSD, OpenBSD, etc
•	4 GB of RAM hardware minimum and a minimum of 250 GB of hard disk space at disposal
One can setup the database with the whole planet data and including meta data, and this has a file size of more than 20 GB.

**Comment by jeffbl** [comment]
Looks like there is already a docker overpass image that was recently updated and has 100k+ pulls:
https://hub.docker.com/r/wiktorn/overpass-api

**Comment by samueljohnsegun148** [comment]
@jeffbl Yes there is. I have checked that as well and it looks very good to me, but I do not just fully know how to take it further from there.   It sounds great to hear from @JRegimbal that Pegasus should be able to take care of the resource requirements. @JRegimbal I hope some of these links provide useful information on OSM/overpass api:  https://python-overpy.readthedocs.io/en/latest/  
https://overpass-api.de/index.html 
https://overpass-api.de/no_frills.html#required, 
https://wiki.openstreetmap.org/wiki/Overpass_API
https://wiki.openstreetmap.org/wiki/Main_Page

**Comment by samueljohnsegun148** [comment]
> @jeffbl It looks good to me, although I admittedly do not know much about OSM/Overpass API. We do currently have close to 800 GB free on pegasus so running it there shouldn't be a problem.

I am currently experimenting with an online Linux server (DigitalOcean).  I'm thinking Pegasus is the production server, so do we have the same/similar resources on the unicorn  server to try things out? Or is it possible that I have ssh or any  appropriate access to the pegasus?

**Comment by jeffbl** [comment]
I believe @JRegimbal is also working on getting something going on docker for this. I am not sure if unicorn has the resources to effectively run this, but if it is exposed publicly on pegasus, then unicorn could query pegasus.

File: .gitignore
Comment (JRegimbal) [inline .gitignore:8]: This should not be here

File: preprocessors/openstreetmap/main.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/main.py]: Still awaiting his final consent.

**Comment by samueljohnsegun148** [comment]
> 

Yes. It's working on the unicorn.

File: preprocessors/openstreetmap/osm_service.py
Comment (JRegimbal) [inline preprocessors/openstreetmap/osm_service.py:535]: Is there a reason you're using your own quicksort rather than python's builtin [timsort](https://en.wikipedia.org/wiki/Timsort)? It doesn't seem to provide any performance advantage based on what I see here.

**Review by JRegimbal — CHANGES_REQUESTED** [review]

File: preprocessors/openstreetmap/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py:535]: Not really, initially I was only having issues with  some of the python inbuilt functions for sorting. I guess because my data is a dictionary type, so was getting this error "**TypeError: '<' not supported between instances of 'dict' and 'dict**'". In order to get a way round it, I decided to solve it my own way. The good news now is that I was able to figure it  out today, and  the sorting can now be done with just a single line of code, using the python built-in sort function. I  will commit right away. Thanks

Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py]: Changes pushed.

**Review by JRegimbal — CHANGES_REQUESTED** [review]
A few minor changes...since none of it requires a major change in functionality, have you been testing with an override on unicorn @samueljohnsegun148?

File: preprocessors/openstreetmap/main.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/main.py:102]: > This is still not in a single variable...

```response = {
                "request_uuid": request_uuid,
                "timestamp": time_stamp,```

Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/main.py:102]:
``` 
 time_stamp = int(get_timestamp())
response = {
                "request_uuid": request_uuid,
                "timestamp": time_stamp,

```
Is it okay like this?

File: preprocessors/openstreetmap/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py]: I've modified it now.

File: preprocessors/openstreetmap/osm_service.py
Comment (samueljohnsegun148) [reply preprocessors/openstreetmap/osm_service.py]: I've modified it now.

**Comment by samueljohnsegun148** [comment]
> 

Yes, I've been using override for the testing on unicorn.

**Comment by samueljohnsegun148** [comment]
> A few minor changes...since none of it requires a major change in functionality, have you been testing with an override on unicorn @samueljohnsegun148?

Yes, I've been using an override for the testing on unicorn.

**Review by JRegimbal — COMMENTED** [review]
Only remaining issue is the name (`ca.mcgill.a11y.image.preprocessor.openstreetmap`) being written out in several places rather than being in a variable. That's not blocking though. If @jeffbl if fine with hard coded fallbacks for right now and configurable fallbacks in a later issue, I'll approve.

**Review by JRegimbal — APPROVED** [review]
Fine as is, some future work will be needed on #472 and refactoring never hurts but this looks good to go. Thanks for all your work @samueljohnsegun148!

**Pull request merged**
